core/client/src/main/java/alluxio/client/block/UfsBlockInStream.java

```diff
@@ -0,0 +1,244 @@
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.nio.ByteBuffer;
+
+import javax.annotation.concurrent.NotThreadSafe;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * This class provides a streaming API to read a UFS block from an Alluxio worker. The data will
+ * be transferred through an Alluxio worker's dataserver to the client.
+ */
+@NotThreadSafe
+public final class UfsBlockInStream extends BufferedBlockInStream implements Locatable {
+  private static final Logger LOG = LoggerFactory.getLogger(UfsBlockReader.class);
+  private static final long UFS_BLOCK_OPEN_RETRY_INTERVAL_MS = Constants.SECOND_MS;
+
+  /** Used to manage closeable resources. */
+  private final Closer mCloser;
+  private final boolean mNoCache;
+  private final boolean mLocal;
+
+  /** Client to communicate with the remote worker. */
+  private final BlockWorkerClient mBlockWorkerClient;
+  /** The file system context which provides block worker clients. */
+  private final FileSystemContext mContext;
+  /** {@link UfsBlockReader} for this instance. */
+  private UfsBlockReader mReader;
+
+  /**
+   * Creates a UFS block in stream.
+   *
+   * @param context the file system context
+   * @param ufsPath the UFS path
+   * @param blockId the block ID
+   * @param blockSize the block size
+   * @param blockStart the start position of the block in the UFS file
+   * @param workerNetAddress the worker network address
+   * @param options the in stream options
+   * @return the {@link UfsBlockInStream} instance
+   * @throws IOException if it fails to create {@link UfsBlockInStream}
+   */
+  public static UfsBlockInStream createUfsBlockInStream(FileSystemContext context, String ufsPath,
+      long blockId, long blockSize, long blockStart, WorkerNetAddress workerNetAddress,
+      InStreamOptions options) throws IOException {
+    Closer closer = Closer.create();
+    try {
+      BlockWorkerClient blockWorkerClient =
+          closer.register(context.createBlockWorkerClient(workerNetAddress));
+      LockBlockResult result;
+      LockBlockOptions lockBlockOptions =
+          LockBlockOptions.defaults().setUfsPath(ufsPath).setOffset(blockStart)
+              .setBlockSize(blockSize).setMaxUfsReadConcurrency(options.getMaxUfsReadConcurrency());
+
+      long timeout = System.currentTimeMillis() + Configuration
+          .getLong(PropertyKey.USER_UFS_BLOCK_OPEN_TIMEOUT_MS);
+      while (true) {
+        try {
+          result = blockWorkerClient.lockBlock(blockId, lockBlockOptions);
+          break;
+        } catch (UfsBlockAccessTokenUnavailableException e) {
+          if (System.currentTimeMillis() >= timeout) {
+            throw e;
+          }
+          LOG.debug(
+              "Failed to acquire a UFS read token because of concurrency for block {} in file {}",
+              blockId, blockId);
+          CommonUtils.sleepMs(UFS_BLOCK_OPEN_RETRY_INTERVAL_MS);
+        }
+      }
```

Comment:
It is a little hard as waitFor doesn't allow the function being waited for to throw any exception. I can add a version of waitFor to accept exception but it is hard to be generic to preserve the exception type. 

Instead, i moved this logic to BlockWorkerClient. let me know if that doesn't look good to you.